File: Examples/intel_examples/examples_core_c/vslc/source/vslssort.c
/*******************************************************************************
* Copyright 2003-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    Sorting data array  Example Program Text
!******************************************************************************/

#include <stdio.h>

#include "mkl.h"
#include "mkl_vsl.h"
#include "errcheck.inc"
#include "generatedata.inc"

#define P       3    /* Number of observations */
#define N       10   /* Task dimension */
#define SEED    1    /* Initial value for stream initialization */

int main()
{
    VSLSSTaskPtr task;
    MKL_INT n;
    MKL_INT p;
    MKL_INT x_storage;
    MKL_INT sorted_x_storage;
    float x[P][N];                    /* Matrix of observations */
    float y[P][N];                    /* Output  matrix for sorted data */
    float lBound = 0.0, rBound = 10;  /* Bounds for uniform generator */
    int errnums = 0;
    int i, j, errcode;
    VSLStreamStatePtr stream;

    /***** Initializing parameters for Summary Statistics task *****/
    n = N;
    p = P;
    x_storage        = VSL_SS_MATRIX_STORAGE_ROWS;
    sorted_x_storage = VSL_SS_MATRIX_STORAGE_ROWS;

    /***** Generate data set using VSL Uniform RNG *****/
    /***** Initialize *****/
    errcode = vslNewStream(&stream, VSL_BRNG_MCG31, SEED);
    CheckVslError(errcode);

    /***** Call RNG *****/
    errcode = vsRngUniform(VSL_RNG_METHOD_UNIFORM_STD, stream, N * P, (float *) x, lBound, rBound);
    CheckVslError(errcode);

    /***** Create Summary Statistics task *****/
    errcode = vslsSSNewTask(&task, &p, &n, &x_storage, (float *) x, 0, 0);
    CheckVslError(errcode);

    /***** Edit task parameters for sorting *****/
    errcode = vslsSSEditTask(task, VSL_SS_ED_SORTED_OBSERV, (float *) y);
    CheckVslError(errcode);

    /***** Edit task parameters for sorting *****/
    errcode = vsliSSEditTask(task, VSL_SS_ED_SORTED_OBSERV_STORAGE, &sorted_x_storage);
    CheckVslError(errcode);

    /***** Sort data using radix method *****/
    errcode = vslsSSCompute(task, VSL_SS_SORTED_OBSERV, VSL_SS_METHOD_RADIX);
    CheckVslError(errcode);

    /***** Check the correctness of sorting *****/
    for( j = 0; j < p; j++ )
    {
        for ( i = 0; i < n - 1; i++ )
        {
            if ( y[j][i] > y[j][i + 1] )
            {
                errnums++;
            }
        }
    }

    /***** Printing results *****/
    printf("Task dimension : %d\n", (int)p);
    printf("Number of observations : %d\n\n", (int)n);

    /***** Printing of the initial matrix *****/
    printf("\n Initial matrix: \n");
    for ( j = 0; j < p; j++ )
    {
        for ( i = 0; i < n; i++ )
        {
            printf("%+.3f   ", x[j][i]);
        }
        printf("\n");
    }
    printf("\n");

    /***** Printing of the sorted matrix *****/
    printf("\n Sorted matrix: \n");
    for ( j = 0; j < p; j++ )
    {
        for ( i = 0; i < n; i++ )
        {
            printf("%+.3f   ", y[j][i]);
        }
        printf("\n");
    }

    /***** Printing summary of the test *****/
    if (errnums == 0)
    {
        printf("\n\n Sorting is correct \n\n");
    }
    else
    {
        printf("\n\n Error: Sorting is incorrect \n\n");
        return 1;
    }

    /***** Delete Summary Statistics task *****/
    errcode = vslSSDeleteTask(&task);
    CheckVslError(errcode);

    MKL_Free_Buffers();

    return 0;
}

File: Examples/intel_examples/examples_core_f/vslf/source/vslssort.f
!===============================================================================
! Copyright 2003-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!    Sorting data array  Example Program Text
!******************************************************************************/

      include 'mkl_vsl.f90'
      include "errcheck.inc"
      include "generatedata.inc"
      include "statchars.inc"

      program MKL_VSL_TEST

      USE MKL_VSL_TYPE
      USE MKL_VSL

      integer,parameter :: DIM = 3      ! Task dimension
      integer,parameter :: NN  = 10     ! Number of observations

      TYPE(VSL_STREAM_STATE) :: stream
      integer,parameter :: SEED   = 1
      integer,parameter :: BRNG   = VSL_BRNG_MCG31
      integer,parameter :: METHOD = VSL_RNG_METHOD_UNIFORM_STD

      TYPE(VSL_SS_TASK) task
      integer p
      integer n
      integer x_storage
      integer y_storage
      real(kind=4) x(DIM, NN)
      real(kind=4) y(DIM, NN)
      real(kind=4) lBound, rBound
      integer i, j
      integer(kind=4) errcode
      integer errnums
      integer task_method
      integer edit_value
      integer(kind=8) task_params

!     ***** Initialize parameters *****
      p           = DIM
      n           = NN
      x_storage   = VSL_SS_MATRIX_STORAGE_COLS
      y_storage   = VSL_SS_MATRIX_STORAGE_COLS
      lBound      = 0.0
      rBound      = 10.0
      task_params = VSL_SS_SORTED_OBSERV
      task_method = VSL_SS_METHOD_RADIX
      edit_value  = VSL_SS_ED_SORTED_OBSERV_STORAGE
      errcode     = 0
      errnums     = 0

!     ***** Generate data set using Uniforn RNG *****
!     ***** Initialize *****
      errcode = vslnewstream( stream, BRNG,  SEED )
      call CheckVslError(errcode)

!     ***** Call RNG *****
      errcode = vsrnguniform(METHOD, stream, n * p, x, lBound, rBound)
      call CheckVslError(errcode)

!     ***** Create Summary Statistics task *****
      errcode = vslsssnewtask( task, p, n, x_storage, x )
      call CheckVslError( errcode )

      errcode = vslsssedittask( task, VSL_SS_ED_SORTED_OBSERV, y )
      call CheckVslError( errcode )

      errcode = vslissedittask( task, edit_value, y_storage )
      call CheckVslError( errcode )

!     ***** Sort data using RADIX method  *****
      errcode = vslssscompute( task, task_params, task_method )
      call CheckVslError( errcode )

!     ***** Check the correctness of sorting *****
      do j = 1, n - 1
        do i = 1, p-1
          if( y(i, j) >= y(i, j + 1) ) then
            errnums = errnums + 1
          end if
        end do
      end do

!     ***** Printing results *****
      print *, ''
      print *, 'Task dimension :        ', p
      print *, 'Number of observations :', n
      print *, ''

!     ***** Printing of the initial matrix *****
      print *,'Initial matrix'
      do j = 1, n
        do i = 1, p
           write(*, 5) x(i,j), ' '
        end do
        print *, ''
      end do


!     ***** Printing of the sorted matrix *****
      print *,'Sorted matrix:'
      do j = 1, n
        do i = 1, p
          write(*, 5) y(i,j), ' '
        end do
        print *, ''
      end do

      print *, ''

!     ***** Printing summary of the test *****
      if( errnums == 0 ) then
        print *, 'Sorting is correct'
      else
        print *, 'Error: Sorting is incorrect'
        stop 1
      end if

!     ***** Delete Summary Statistics task *****
      errcode = vslssdeletetask( task )
      call CheckVslError( errcode )

      call MKL_FREE_BUFFERS()

5     format (F6.3,A,$)

      end
